src/details/hooks.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+class kaleidoscope_;
+
+namespace kaleidoscope {
+
+class KeyEvent;
+union Key;
+
+// The reason why the hook rooting entry point functions live within
+// class Hooks and not directly within a namespace is that we want
+// to restrict who is alowed to trigger hooks. This is mainly to prevent
+// user code from calling hook methods.
+//
+// A note to maintainers: When you add new hooks that are supposed to
+// be called from other places than the friend classes and functions listed
+// below, just add a suitable friend declaration.
+
+class Hooks {
```

Comment:
Any compilation units that are supposed to call hook functions only include `details/hooks.h` without a need to know anything about hook routing. Header `hooks.h` has been moved to `details` to make sure that it is not part of the plugin interface.